### ISSUE TITLE
Improve cli for better scriptability

### ISSUE THREAD

**Issue by cjellick** [rancher/rancher#15692]
The CLI needs refined to make it easier to write a script to bootstrap a cluster. overall goal is to setup rancher with single node, login with CLI, spin up cluster and then set the context into the default project of the new cluster. with consistent error messages. The following details are copied in from another conversation, so they arent in perfect issue reporting form.

single node rancher, create user, login to cli (get warning that there is no projects), use `rancher up`  to create a GKE cluster (using a cluster.yml from a previous setup)

now they are polling to try and figure out when the cluster is active, `rancher cluster ls` - there are some statuses/columns missing while provisioning is happening and there is no warning about having no project set or there is an error? something like that
when the cluster is finally active, you still haven't set a default project and there is no definitive command to figure out how to set the default project. they currently use `rancher project ls` but there are 2 projects (not always in the same order). they select one, check to see if it's default, if not, they switch again.
overall goal is to setup rancher with single node, login with CLI, spin up cluster and then set the context into the default project of the new cluster. with consistent error messages.

FYI, this is the script of what they are doing today:

```
# Login with the APITOKEN as key for rancher cli without any project context 
echo "yes" | $RANCHER_CLI_HOME/rancher login "$BASE_URL"  -t $APITOKEN

# Needed to prevent no project set errors
$RANCHER_CLI_HOME/rancher context switch

# Bring up cluster
$RANCHER_CLI_HOME/rancher up -f cluster.yml

# During provision of cluster, a Default project gets created which we will use as context

project_id=$($RANCHER_CLI_HOME/rancher context switch < /dev/null | grep Default | awk ' {print $3}')
while [ -z $project_id ]; do
        echo "Default project not yet created .."
        sleep 15
        project_id=$(./rancher context switch < /dev/null | grep Default | awk ' {print $3}')
done

echo "Default project created with project_id : $project_id Logging in with context set to this project id"
echo "yes" | $RANCHER_CLI_HOME/rancher login "$BASE_URL"  -t $APITOKEN --context "$project_id"

# Once logged in with the context, all commands work as expected. 
state=$($RANCHER_CLI_HOME/rancher clusters ls | grep cluster1 | awk '{print $3}')

while [ "$state" != "active" ]; do
        echo "State is $state, Waiting till it becomes Active"
        sleep 120;
        state=$($RANCHER_CLI_HOME/rancher clusters ls | grep cluster1 | awk '{print $3}')
        
done
```

```
Logging in first without project, create a cluster and wait for default project and then relogin with the default project_id in the context eliminates all errors and warnings. Wihout this, the rancher clusters ls, does not have data for the first column, so I have to $2 instead for the cluster state. Please let me know this approach is fine or if there is an alternate approach recommended. For the API token, I am following the steps in https://medium.com/@superseb/adding-custom-nodes-to-your-kubernetes-cluster-in-rancher-2-0-tech-preview-2-89cf4f55808a and those steps seem to be working fine.```

**Comment by dramich** [comment]
```
# Login with the APITOKEN as key for rancher cli without any project context 
$RANCHER_CLI_HOME/rancher login "$BASE_URL"  -t $APITOKEN --skip-verify

# Bring up cluster
$RANCHER_CLI_HOME/rancher up -f cluster.yml

timeout=120
declare -i count=0
# Use inspect to know when the cluster is on the server so we can watch it
# If the warnings are a concern the context switch could be after this inspect
while $RANCHER_CLI_HOME/rancher inspect $cluster_name
do
    if [ $count -gt $timeout ]
    then
        echo "Timout reached, check the server to see if the cluster is there"
        exit 1
    fi
    sleep 0.5
    count+=1
done

# Use the cluster name from your cluster.yml, timeout is whatever is needed, defaults to 120 seconds
eval "$RANCHER_CLI_HOME/rancher wait $cluster_name --timeout 600"

# Context switch will default to the Default project if there are only 1 or 2
# projects and they are Default and System. This assumes there is only 1 cluster
# in the server. If there is more than one you can pass the projectID to context switch
$RANCHER_CLI_HOME/rancher context switch
```

**Comment by dramich** [comment]
`login` now has a `skip-verify` flag if you are using a self signed cert. Enabling the flag should skip the `do you trust this cert` check and just continue the login. 

`login` now defaults to the `Default` project if there are two projects in the cluster and they are `System` and `Default`. This is true for `context switch` as well and should be tested. 

New `wait` command which holds until a cluster is in an active state based on a timeout.

**Comment by soumyalj** [comment]
@cjellick : The `wait ` command is added as a separate command in the CLI and we wait for only "active" state. Shouldn't it be an option as part of `create` command for any resource where in we wait for the specific resource to go to active state. In the current implementation its an extra step for users to create and then wait.

**Comment by cjellick** [comment]
We'll defer to monday. Will talk through the UX with @ibuildthecloud and @vincent99 then, but this is the pre-established pattern for wait in previous iterations of the cli

**Comment by vincent99** [comment]
The while in the new script form @dramich is pretty weird, shouldn't `up` not exit until the resources in the yams file are created (and thus get-able immediately after)?

Not sure how an "inline" `--wait` would work on things that create multiple resources.  Or that usage of it would end up being any shorter in a real-world script.  You'd need a different exit code to differentiate "created but didn't go active" from "didn't create in the first place", and then the script needs to know what specific numbers those are instead of just checking if create returns 0 and then wait returns 0.

**Comment by cjellick** [comment]
@soumyalj we're going to go with this paradigm, no inline wait.

**Comment by soumyalj** [comment]
Tested with rancher cli  v2.2.0-rc5. As mentioned above, login --skip-verify and wait has been introduced .
1. login --skip-verify is functional when using default self signed certs .Using this flag skips the `do you trust this cert` check and just continues the login as expected.

2. The wait command fails with the error as below. 

```
rancher-v2.2.0-rc5 soumya$ ./rancher cluster ls
CURRENT   ID        STATE     NAME            PROVIDER                    NODES     CPU       RAM             PODS
          c-bf68j   active    rosclusteragn   Rancher Kubernetes Engine   3         1.34/6    0.64/11.27 GB   20/330
          c-dldts   pending   dotestnew       Unknown                     0         0.00/0    0.00/0.00 GB    0/0
*         c-q52fj   active    dotest          Rancher Kubernetes Engine   1         0.54/2    0.14/3.76 GB    9/110
rancher-v2.2.0-rc5 soumya$ ./rancher wait dotestnew --timeout 60
FATA[0010] unknown resource type                        
rancher-v2.2.0-rc5 soumya$ ./rancher wait --timeout 60 c-dldts
FATA[0010] unknown resource type                        

```

**Comment by dramich** [comment]
Wait is fixed in v2.2.0-rc6

**Comment by soumyalj** [comment]
Tested with CLI v2.2.0-rc6 on master build from Feb 12.
The following issues are seen in the wait command:
1. During cluster creation , in the initial stages, when the node is being created, wait command does not display a correct message. It displays nil `%!s(<nil>` while there is actually a message available in the API. 
```
soumyas-MacBook-Pro:rancher-v2.2.0-rc6 soumya$ ./rancher wait doclustersing --timeout 5
FATA[0005] Timeout reached cluster:c-fzhhk failed: %!s(<nil>) 

```
API message below:
```
"state": "provisioning",
"transitioning": "yes",
"transitioningMessage": "waiting for soumyakubectltest7 to finish provisioning",

```
2. When a wait command is issued for an active cluster,  it gives out an error as below:
```
soumyas-MacBook-Pro:rancher-v2.2.0-rc6 soumya$ ./rancher wait doclustersing --timeout 10
FATA[0010] Timeout reached cluster:c-fzhhk failed: %!s(<nil>) 

```

**Comment by soumyalj** [comment]
Tested with rancher version v2.2.0-rc7. 
./rancher cluster up fails with the below parsing error.
```
soumya$ ./rancher cluster up --file cluster.yml 
strconv.ParseFloat: parsing "1329952Ki": invalid syntax
```
Since this fails, the wait command cannot be tested

**Comment by dramich** [comment]
The error is from an issue in the yaml file. `1329952Ki` has letters, it's expecting just a number. Wait can be tested without using the `up` command. Create a cluster in the UI then wait on that cluster through the cli.

**Comment by soumyalj** [comment]
A separate issue is filed for the parsing error seen.
Tested with rancher version v2.2.0-rc7.
1. During cluster creation , in the initial stages, when the node is being created, wait command  displays a message as expected without any errors. Example below:
```
rancher-v2.2.0-rc7 soumya$ ./rancher up --file clustertest.yml 
rancher-v2.2.0-rc7 soumya$ ./rancher wait --timeout 5 dosingnodenew101
FATA[0005] Timeout reached cluster:c-j8fvh transitioningMessage: waiting for soumyadosingnode9 to finish provisioning 

```
2. When a wait command is issued for an active cluster,  it just returns as expected. 

```
rancher-v2.2.0-rc7 soumya$ ./rancher wait --timeout 5 dosingnodenew100
soumyas-MacBook-Pro:rancher-v2.2.0-rc7 soumya$ 
```
